### PR TITLE
[Precision Depth Alignment] Fix paddle.nn.functional.affine_grid forward and backward gradient calculation

### PR DESCRIPTION
**Pull request by zhengshengning**
### PR Category

Operator Mechanism

### PR Types

New features 

### Description

**修改内容：**
1. 分别对4D和5D手写了对应的cuda kernel实现，由于新增的Cuda实现性能相比之前要差，所以通过FLAGS_torch_compatible_kernel 来控制是否启用精度对齐方案，默认是关闭的；
2. 增加单元测试

**前向传播**
1. 生成基础网格：
   - 4D: 为每个(H, W)位置生成归一化坐标(x, y, 1)
   - 5D: 为每个(D, H, W)位置生成归一化坐标(x, y, z, 1)

2. 仿射变换：
   - 4D: `[N, H×W, 3] × [N, 3, 2] = [N, H×W, 2]`
   - 5D: `[N, D×H×W, 4] × [N, 4, 3] = [N, D×H×W, 3]`

**反向传播**
1. 梯度计算：
   - 4D: `[N, 3, H×W] × [N, H×W, 2] = [N, 3, 2]` → 转置为 [N, 2, 3]
   - 5D: `[N, 4, D×H×W] × [N, D×H×W, 3] = [N, 4, 3]` → 转置为 [N, 3, 4]

两者遵循相同的计算模式：生成基础网格 → 应用仿射变换 → 输出变换后的网格坐标。

**测试结果：**
1.  测试集的 paddle.nn.functional.affine_grid 的4D case 全部对齐（10个）；
4. 新增 10 个 5D的测试case全部通过（10个）；

Pcard-67164

### CODE REVIEW

**Review by wanghuancoder — APPROVED** [review]
LGTM
